src/Workspaces/Core/Portable/Utilities/CancellationTokenExtensions.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Shared.Utilities
+{
+    /// <summary>
+    /// Extensions to <see cref="CancellationToken"/>.
+    /// </summary>
+    internal static class CancellationTokenExtensions
+    {
+        /// <summary>
+        /// Creates a new <see cref="CancellationToken"/> that is canceled when any of a set of other tokens are canceled.
+        /// </summary>
+        /// <param name="original">The first token.</param>
+        /// <param name="other">The second token.</param>
+        /// <returns>A struct that contains the combined <see cref="CancellationToken"/> and a means to release memory when you're done using it.</returns>
+        public static CombinedCancellationToken CombineWith(this CancellationToken original, CancellationToken other)
+        {
+            if (original.IsCancellationRequested || !other.CanBeCanceled)
+            {
+                return new CombinedCancellationToken(original);
+            }
+
+            if (other.IsCancellationRequested || !original.CanBeCanceled)
+            {
+                return new CombinedCancellationToken(other);
+            }
+
+            // This is the most expensive path to take since it involves allocating memory and requiring disposal.
+            // Before this point we've checked every condition that would allow us to avoid it.
+            return new CombinedCancellationToken(CancellationTokenSource.CreateLinkedTokenSource(original, other));
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="CancellationToken"/> that is canceled when any of a set of other tokens are canceled.
+        /// </summary>
+        /// <param name="original">The first token.</param>
+        /// <param name="others">The additional tokens.</param>
+        /// <returns>A struct that contains the combined <see cref="CancellationToken"/> and a means to release memory when you're done using it.</returns>
+        public static CombinedCancellationToken CombineWith(this CancellationToken original, params CancellationToken[] others)
+        {
+            Contract.ThrowIfNull(others, nameof(others));
+
+            if (original.IsCancellationRequested)
+            {
+                return new CombinedCancellationToken(original);
+            }
```

Comment:
this seems to have different semantics from above.

Reply:
Can you describe an input pair (CancellationToken a, CancellationToken b) for which the following are not the same?

* `CombineWith(a, b)`
* `CombineWith(a, new[] { b })`

Reply:
Retracted.  Misread things

Reply:
FWIW I had the same misread. :smile: